ast/src/parsed/pattern.rs

```diff
@@ -0,0 +1,432 @@
+
+                        let mut new_patterns = Vec::with_capacity(remaining_part.len() + 1);
+                        new_patterns.push(unspecialized);
+                        new_patterns.extend_from_slice(remaining_part);
+                        Some(PatternTuple {
+                            patterns: new_patterns,
+                        })
+                    } else {
+                        None
+                    }
+                }
+
+                Pattern::Enum(path1, _) => Some(PatternTuple {
+                    patterns: vec![Pattern::Enum(path1.clone(), Some(tuple.patterns))],
+                }),
+                _ => None,
+            }
+        }
+    }
+}
+
+impl Default for PatternTuple {
+    fn default() -> Self {
+        PatternTuple { patterns: vec![] }
+    }
+}
+
+#[derive(
+    Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema, Hash,
+)]
+pub enum Pattern {
+    CatchAll, // "_", matches a single value
+    Ellipsis, // "..", matches a series of values, only valid inside array patterns
+    #[schemars(skip)]
+    Number(BigInt),
+    String(String),
+    Tuple(Vec<Pattern>),
+    Array(Vec<Pattern>),
+    // A pattern that binds a variable. Variable references are parsed as
+    // Enum and are then re-mapped to Variable if they do not reference
+    // an enum variant.
+    Variable(String),
+    Enum(SymbolPath, Option<Vec<Pattern>>),
+}
+
+impl Pattern {
+    /// Returns an iterator over all variables in this pattern.
+    pub fn variables(&self) -> Box<dyn Iterator<Item = &String> + '_> {
+        match self {
+            Pattern::Variable(v) => Box::new(once(v)),
+            _ => Box::new(self.children().flat_map(|p| p.variables())),
+        }
+    }
+
+    /// Return true if the pattern is irrefutable, i.e. matches all possible values of its type.
+    pub fn is_irrefutable(&self) -> bool {
+        // TODO This should also return true for enum variants that are the only ones in the enum
+        match self {
+            Pattern::Ellipsis => unreachable!(),
+            Pattern::CatchAll | Pattern::Variable(_) => true,
+            Pattern::Number(_) | Pattern::String(_) | Pattern::Enum(_, _) => false,
+            Pattern::Array(items) => {
+                // Only "[..]"" is irrefutable
+                items == &vec![Pattern::Ellipsis]
+            }
+            Pattern::Tuple(p) => p.iter().all(|p| p.is_irrefutable()),
+        }
+    }
+
+    /// Specialize a pattern based on a "constructor" pattern passed as a parameter.
+    /// Based on https://doc.rust-lang.org/nightly/nightly-rustc/rustc_pattern_analysis/usefulness/index.html#specialization.
+    /// If the constructor is shared with the pattern, specialize the pattern removing
+    /// "one layer" (the constructor) and returning internal patterns.
+    /// Return None if the pattern can't be specialized using the constructor.
```

Comment:
Ok, I think this part I understand now. Let me rephrase it in my own words and then we can see how we want to combine the docs.
```suggestion
    /// We are only interested in the type constructor part of `constructor` and ignore the inner patterns and fields.
    /// If the constructor does not match the pattern, returns None.
    /// If it matches, returns a vector of corresponding constructor patterns for the inner fields of the pattern.
```

And then maybe add some formal statement like on https://doc.rust-lang.org/nightly/nightly-rustc/rustc_pattern_analysis/usefulness/index.html#specialization

Reply:
Ok it seems I still have not understood it :)
The returned patterns correspond to the sub-items of the constructor, not of the pattern, right?